src/scheduler/simulate.c

```diff
@@ -1284,9 +1284,16 @@ find_event_ptr(timed_event *ote, server_info *nsinfo)
 		case TIMED_RUN_EVENT:
 		case TIMED_END_EVENT:
 			oep = (resource_resv *) ote->event_ptr;
-			event_ptr =
-				find_resource_resv_by_time(nsinfo->all_resresv,
-				oep->name, oep->start);
+			if (oep->is_resv)
+				event_ptr =
+					find_resource_resv_by_time(nsinfo->all_resresv,
+					oep->name, oep->start);
+			else
+				/* In case of jobs there can be only one occurance of job in
+				 * all_resresv list, so no need to search using start time of job
+				 */
+				event_ptr = find_resource_resv_by_indrank(nsinfo->all_resresv, 
```

Comment:
search by rank is O(N) operation and search by index is O(1).
It was earlier search by rank but we changed it to introduce index.

Reply:
I kind of feel that it'll be better (and more OO) if we have just 1 generic function that takes as arguments "name", "rank" and "index", it first searches by index if it's not negative, then by rank if it's not negative, and then, finally, by name. If there's a 'time' argument then it can also check for that. I'd personally prefer that over multiple functions, but I'll leave it to you to decide if you wanna refactor the code or not.

Reply:
I agree with you, maybe we can have a single generic function to find the resource resv.
I don't want to change that with this pull request though, it will expand the scope of this work.